core/pta/attestation.c

```diff
@@ -0,0 +1,761 @@
+	p += sizeof(val);
+	max_sz -= sizeof(val);
+	if (max_sz < sz)
+		return 0;
+
+	res = crypto_bignum_bin2bn(p, sz, bn);
+	if (res)
+		return 0;
+
+	return sizeof(val) + sz;
+}
+
+static TEE_Result deserialize_key(uint8_t *buf, size_t buf_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *p = buf;
+	size_t sz = 0;
+
+	res = allocate_key();
+	if (res)
+		goto out;
+
+	sz = deserialize_bignum(p, buf_sz, key->e);
+	if (!sz)
+		goto out;
+	p += sz;
+	buf_sz -= sz;
+	sz = deserialize_bignum(p, buf_sz, key->d);
+	if (!sz)
+		goto out;
+	p += sz;
+	buf_sz -= sz;
+	sz = deserialize_bignum(p, buf_sz, key->n);
+	if (!sz)
+		goto out;
+	buf_sz -= sz;
+	assert(!buf_sz);
+out:
+	return res;
+}
+
+static TEE_Result sec_storage_obj_read(TEE_UUID *uuid, uint32_t storage_id,
+				       const uint8_t *obj_id,
+				       size_t obj_id_len,
+				       uint8_t *data, size_t *len,
+				       size_t offset, uint32_t flags)
+{
+	const struct tee_file_operations *fops = NULL;
+	TEE_Result res = TEE_ERROR_BAD_STATE;
+	struct tee_file_handle *fh = NULL;
+	struct tee_pobj *po = NULL;
+	size_t file_size = 0;
+	size_t read_len = 0;
+
+	fops = tee_svc_storage_file_ops(storage_id);
+	if (!fops)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (obj_id_len > TEE_OBJECT_ID_MAX_LEN)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = tee_pobj_get(uuid, (void *)obj_id, obj_id_len, flags, false, fops,
+			   &po);
+	if (res)
+		return res;
+
+	res = po->fops->open(po, &file_size, &fh);
+	if (res)
+		goto out;
+
+	read_len = *len;
+	res = po->fops->read(fh, offset, data, &read_len);
+	if (res == TEE_ERROR_CORRUPT_OBJECT) {
+		EMSG("Object corrupt");
+		po->fops->remove(po);
+	} else if (!res) {
+		*len = read_len;
+	}
+
+	po->fops->close(&fh);
+out:
+	tee_pobj_release(po);
+
+	return res;
+}
+
+static TEE_Result sec_storage_obj_write(TEE_UUID *uuid, uint32_t storage_id,
+					const uint8_t *obj_id,
+					size_t obj_id_len,
+					const uint8_t *data, size_t len,
+					size_t offset, uint32_t flags)
+
+{
+	const struct tee_file_operations *fops = NULL;
+	struct tee_file_handle *fh = NULL;
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_pobj *po = NULL;
+
+	fops = tee_svc_storage_file_ops(storage_id);
+	if (!fops)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (obj_id_len > TEE_OBJECT_ID_MAX_LEN)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = tee_pobj_get(uuid, (void *)obj_id, obj_id_len, flags, false,
+			   fops, &po);
+	if (res)
+		return res;
+
+	res = po->fops->open(po, NULL, &fh);
+	if (res == TEE_ERROR_ITEM_NOT_FOUND)
+		res = po->fops->create(po, false, NULL, 0, NULL, 0, NULL, 0,
+				       &fh);
+	if (!res) {
+		res = po->fops->write(fh, offset, data, len);
+		po->fops->close(&fh);
+	}
+
+	tee_pobj_release(po);
+
+	return res;
+}
+
+static TEE_Result load_key(void)
+{
+	/*
+	 * Serialized key pair is 3 bignums (e, p and n) plus their sizes
+	 * encoded as uint16_t. e is 65537 so it needs only 3 bytes.
+	 */
+	size_t size = 3 + 2 * MAX_KEY_SIZE / 8 + 3 * sizeof(uint16_t);
```

Comment:
- I used `MAX_KEY_SIZE` rather than `CFG_ATTESTATION_PTA_KEY_SIZE` because I thought maybe we should be able to read a larger key that might have been generated and saved by a prior version of the TEE.
- I will move the formula to a macro (or even better a static function)
- I will also allocate the buffer within `init_key()` as suggested.

Reply:
Makes sense to me.